core/src/main/java/org/elasticsearch/index/shard/ShardSplittingQuery.java

```diff
@@ -0,0 +1,249 @@
+                                }
+                            });
+                        }
+                    }
+                }
+                return new ConstantScoreScorer(this, score(), new BitSetIterator(bitSet, bitSet.length()));
+            }
+        };
+    }
+
+    @Override
+    public String toString(String field) {
+        return "shard_splitting_query";
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        return sameClassAs(o);
+    }
+
+    @Override
+    public int hashCode() {
+        return classHash();
+    }
+
+    private static void findSplitDocs(String idField, Predicate<BytesRef> includeInShard,
+                                      LeafReader leafReader, IntConsumer consumer) throws IOException {
+        Terms terms = leafReader.terms(idField);
+        TermsEnum iterator = terms.iterator();
+        BytesRef idTerm;
+        PostingsEnum postingsEnum = null;
+        while ((idTerm = iterator.next()) != null) {
+            if (includeInShard.test(idTerm) == false) {
+                postingsEnum = iterator.postings(postingsEnum);
+                int doc;
+                while ((doc = postingsEnum.nextDoc()) != DocIdSetIterator.NO_MORE_DOCS) {
+                    consumer.accept(doc);
+                }
+            }
+        }
+    }
+
+    private static final class Visitor extends StoredFieldVisitor {
+        int leftToVisit = 2;
+        final BytesRef spare = new BytesRef();
+        String routing;
+        String id;
+
+        void reset() {
+            routing = id = null;
+            leftToVisit = 2;
+        }
+
+        @Override
+        public void binaryField(FieldInfo fieldInfo, byte[] value) throws IOException {
+            switch (fieldInfo.name) {
+                case IdFieldMapper.NAME:
+                    id = Uid.decodeId(value);
+                    break;
+                default:
+                    throw new IllegalStateException("Unexpected field: " + fieldInfo.name);
+            }
+        }
+
+        @Override
+        public void stringField(FieldInfo fieldInfo, byte[] value) throws IOException {
+            spare.bytes = value;
+            spare.offset = 0;
+            spare.length = value.length;
+            switch (fieldInfo.name) {
+                case RoutingFieldMapper.NAME:
+                    routing = spare.utf8ToString();
+                    break;
+                default:
+                    throw new IllegalStateException("Unexpected field: " + fieldInfo.name);
+            }
+        }
+
+        @Override
+        public Status needsField(FieldInfo fieldInfo) throws IOException {
+            switch (fieldInfo.name) {
+                case IdFieldMapper.NAME:
+                case RoutingFieldMapper.NAME:
+                    leftToVisit--;
+                    return Status.YES;
+                default:
+                    return leftToVisit == 0 ? Status.STOP : Status.NO;
+
+            }
+        }
```

Comment:
Do you need to support indices created in 5.x? If yes you will need to read the _uid field too (which is a string field).

Reply:
FWIW indices will have either _id or __uid but never both.

Reply:
yeah no we dont' suppport 5.x I will add comments